Create new default theme to match GTK again

twa022 opened the pull request:
This creates a new default theme that tries to dynamically match the GTK theme (as was done in versions before 22.04.0) but using a new method of overriding values in the provided `mate-hud` rofi theme.
The theme is still configurable through the `org.mate.hud.rofi-theme` gsettings key. Only when `mate-hud` or `mate-hud-rounded` are selected does the program attempt to do any theme color matching. If another theme is selected it leaves it be.

A couple more things I think could be implemented, let me know what you think:

- The program currently overrides the theme to position the HUD at the top left (north west) corner of the current window). For people using RTL languages should it override it to the top right (north east)? Is there a better was to check that than comparing  `LANG` environment variable to a list of known RTL locales?
- Include an option to anchor the hud to the screen rather than the window? Wasn't that the default Unity implementation? I use an appmenu plugin in the top panel and think it kind of makes sense to show the HUD below that (or at least provide the option).

twa022 commented:
@flexiondotorg @vkareh Let me know what you think (based on our discussion from #53). I tried to keep the default theme simple, but if you want something fancier I could try and jazz it up. Tested with yaru-*, numix, skeuos-*, nordic, adwaita

vkareh commented:
I'm having several issues with this:

1. The docu theme is now misplaced
![Screenshot at 2022-03-28 15-23-44](https://user-images.githubusercontent.com/259780/160471848-43ffff09-aeec-454f-8623-b76544d851fb.png)

2. The mate-hud theme is too narrow on some windows, cutting off the text
![Screenshot at 2022-03-28 15-24-45](https://user-images.githubusercontent.com/259780/160471917-f9d86dc4-0884-4492-b1f6-0f8f0f3ff970.png)

3. I wonder if the theme itself should be named 'gtk' or 'default' to make it clear that it uses the current default

twa022 commented:
@vkareh I think I fixed the issues you pointed out

1. docu anchored properly
![](https://user-images.githubusercontent.com/652903/160518661-65fdab01-7672-4051-8ae9-e7ea7d1238d1.png)
2. forced default `mate-hud` and `mate-hud-rounded` to 400px width so it's not too small on thin windows. Rofi is supposed to support a max() function [link](https://man.archlinux.org/man/rofi-theme.5#Calculating_sizes) I wanted to do something like max( 400px, 30% ) but it would not parse properly.
![](https://user-images.githubusercontent.com/652903/160519338-615d8c26-6201-4cea-8071-9c250daa9acc.png)
![](https://user-images.githubusercontent.com/652903/160519356-96a49856-5822-4c19-b461-fcb472cd8905.png)

3. The themes are incomplete (certain properties only set by mate-hud at runtime) so shouldn't be used outside of mate-hud so that's why I thought to name them as such.

vkareh commented on usr/lib/mate-hud/mate-hud:
```diff
@@ -174,21 +190,59 @@ def get_menu(menuKeys):
+           '-sync', # withhold display until menu entries are ready
+           '-monitor', '-2', # show in the current application
+           '-theme', rofi_theme,
+           '-theme-str', 'window { location: north west; ' + \
```
Why not provide this as part of the rasi theme itself, and rely on `-location 1` for all other windows?

twa022 replied:
I'm pretty sure the location switch is deprecated [link](https://man.archlinux.org/man/community/rofi/rofi.1.en#Layout_and_Theming) which is why I switched to the new theme format way of doing it.

vkareh commented:
Another quirk (and this happens in `master` as well) is that with compositor enabled, non-maximised resizable windows will show the rofi window offset by the invisible border (10px) in both top and left. It's visually disturbing, but it probably belongs in another PR

twa022 commented:
I could set the margin in the default themes to 0 and then it would position truly at the top left corner. That margin looked good on my theme placing the HUD below the titlebar but since that titlebar height varies theme to theme I see how it could seem inconsistent.

flexiondotorg commented:
Thank you very much for working on the @twa022 🥰

I'll get this a proper test this evening. One question, is it possible to have fonts fall back? Ideally, I'd like to use the Ubuntu font if installed and fall back to Sans Regular.

twa022 commented:
@flexiondotorg

I hadn't looked into the fonts yet, but just did now. I updated the theme overrides for the `mate-hud` and `mate-hud-rounded` themes to include setting the font to match the interface font (the old -font switch wasn't having any effect so I removed it). As is it only overrides the font to match the interface when one of those two themes are selected (same as the other overrides other than position).

flexiondotorg commented:
I'll merge this as is, since it is a huge improvement of the `docu` theme. If you have any further tweak or improvements, please do raise a pull request and I'll ask that if @vkareh has any additional feedback to raise an issue here and tag @twa022 :+1:
